File: GRM-Java/app/src/main/java/com/tartagliaeg/grmjava/datasources/GithubRepo.java
package com.tartagliaeg.grmjava.datasources;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;

import com.google.gson.annotations.SerializedName;

@Entity(tableName = "github_repo")
public class GithubRepo {
  @PrimaryKey
  private final long id;

  @ColumnInfo(name = "name")
  private final String name;

  @ColumnInfo(name = "owner_name")
  private final String ownerName;

  @ColumnInfo(name = "url")
  @SerializedName("html_url")
  private final String url;

  @Ignore
  @SerializedName("owner")
  private final Owner owner;


  public GithubRepo(long id, String name, String url, Owner owner) {
    this.id = id;
    this.name = name;
    this.url = url;
    this.owner = owner;
    this.ownerName = owner.name;
  }

  public GithubRepo(long id, String name, String ownerName, String url) {
    this.id = id;
    this.name = name;
    this.ownerName = ownerName;
    this.owner = new Owner(ownerName);
    this.url = url;
  }

  public long getId() {
    return id;
  }

  public String getOwnerName() {
    return ownerName != null
      ? ownerName
      : owner.name;
  }

  public String getName() {
    return name;
  }

  public String getUrl() {
    return url;
  }


  public static class Owner {
    @SerializedName("login")
    public final String name;

    public Owner(String name) {
      this.name = name;
    }
  }
}

File: GRM-Swift/GRM-Swift/DataSources/GithubRepo.swift
//
//  GithubRepo.swift
//  GRM-Swift
//
//

import Foundation
import RealmSwift

class GithubRepo: Object {
    @objc dynamic var id: Int64 = 0;
    @objc dynamic var name: String = "";
    @objc dynamic var ownerName: String = "";
    @objc dynamic var url: String = "";
    
    convenience init (id: Int64, name: String, ownerName: String, url: String) {
        self.init();
        
        self.id = id;
        self.name = name;
        self.ownerName = ownerName;
        self.url = url;
    }
    
    convenience init (data: Dictionary<String, Any>) {
        self.init();
        
        self.id = (data["id"] ?? 0) as! Int64;
        self.name = (data["name"] ?? "") as! String;
        self.ownerName = ((data["owner"] as! Dictionary<String, Any>)["login"] ?? "") as! String;
        self.url = (data["html_url"] ?? "") as! String;
    }
    
    override static func primaryKey() -> String? {
        return "id"
    }
    
}
